libp2p-swarm/src/connection_reuse.rs

```diff
@@ -66,31 +62,49 @@ pub struct ConnectionReuse<T, C>
 where
 	T: Transport,
 	C: ConnectionUpgrade<T::RawConn>,
+	C::Output: StreamMuxer,
 {
 	// Underlying transport and connection upgrade for when we need to dial or listen.
 	inner: UpgradedNode<T, C>,
+
+	// Struct shared between most of the types that are part of the `ConnectionReuse`
+	// infrastructure.
 	shared: Arc<Mutex<Shared<C::Output>>>,
 }
 
-struct Shared<O> {
-	// List of futures to dialed connections.
-	incoming: Vec<Box<Stream<Item = (O, Multiaddr), Error = future::SharedError<Mutex<Option<IoError>>>>>>,
-	// Tasks to signal when an element is added to `incoming`. Only used when `incoming` is empty.
-	to_signal: Vec<task::Task>,
+struct Shared<M> where M: StreamMuxer {
+	// List of active muxers.
+	active_connections: FnvHashMap<Multiaddr, M>,
```

Comment:
`fnv` is faster for small keys. do you know what is the threshold? when the key is considered to be 'short' and when it is not 'short' any more? I'm just curious

Reply:
I found this (old) link: http://cglab.ca/~abeinges/blah/hash-rs/
Most the time the multiaddr will be something like `/ip4/a.b.c.d/tcp/e` which is 8 bytes. Fnv may indeed be more appropriate here.